api/src/com/cloud/agent/api/storage/OVFHelper.java

```diff
@@ -0,0 +1,333 @@
+            NodeList disks = doc.getElementsByTagName("Disk");
+            NodeList files = doc.getElementsByTagName("File");
+            NodeList items = doc.getElementsByTagName("Item");
+            boolean toggle = true;
+            for (int j = 0; j < files.getLength(); j++) {
+                Element file = (Element)files.item(j);
+                OVFFile of = new OVFFile();
+                of._href = file.getAttribute("ovf:href");
+                if (of._href.endsWith("vmdk") || of._href.endsWith("iso")) {
+                    s_logger.info("MDOVA getOVFVolumeInfo File href = " + of._href);
+                    of._id = file.getAttribute("ovf:id");
+                    s_logger.info("MDOVA getOVFVolumeInfo File Id = " + of._id);
+                    String size = file.getAttribute("ovf:size");
+                    if (size != null && !size.isEmpty()) {
+                        of._size = Long.parseLong(size);
+                    } else {
+                        String dataDiskPath = ovfFile.getParent() + File.separator + of._href;
+                        File this_file = new File(dataDiskPath);
+                        of._size = this_file.length();
+                    }
+                    of._iso = of._href.endsWith("iso");
+                    if (toggle && !of._iso) {
+                        of._bootable = true;
+                        toggle = !toggle;
+                    }
+                    vf.add(of);
+                }
+            }
+            for (int i = 0; i < disks.getLength(); i++) {
+                Element disk = (Element)disks.item(i);
+                OVFDisk od = new OVFDisk();
+                String virtualSize = disk.getAttribute("ovf:capacity");
+                if (virtualSize == null || virtualSize.isEmpty()) {
+                    od._capacity = 0L;
+                } else {
+                    od._capacity = Long.parseLong(virtualSize);
+                }
+                String allocationUnits = disk.getAttribute("ovf:capacityAllocationUnits");
+                od._diskId = disk.getAttribute("ovf:diskId");
+                s_logger.info("MDOVA getOVFVolumeInfo Disk ovf:diskId  = " + od._diskId);
+                od._fileRef = disk.getAttribute("ovf:fileRef");
+                s_logger.info("MDOVA getOVFVolumeInfo Disk ovf:fileRef  = " + od._fileRef);
+                od._populatedSize = Long.parseLong(disk.getAttribute("ovf:populatedSize") == null ? "0" : disk.getAttribute("ovf:populatedSize"));
+                s_logger.info("MDOVA getOVFVolumeInfo Disk _populatedSize  = " + od._populatedSize);
+
+                if ((od._capacity != 0) && (allocationUnits != null)) {
+
+                    long units = 1;
+                    if (allocationUnits.equalsIgnoreCase("KB") || allocationUnits.equalsIgnoreCase("KiloBytes") || allocationUnits.equalsIgnoreCase("byte * 2^10")) {
+                        units = 1024;
```

Comment:
Don't we have these "magic" numbers already in a variable? If not, What about creating a enum class for this type of thing

BTW: `2^10 != 1 KB (KiloBytes)` ;
`1 KB (KiloBytes) == 1000`
`1 KiB (Kibibyte) == 1024`

Do you know the real unit used here?

Reply:
Thanks for this finding. I have checked OVF spec on http://www.dmtf.org/sites/default/files/standards/documents/DSP0243_1.1.0.pdf and found this:

````
The capacity of a virtual disk shall be specified by the ovf:capacity attribute with an xs:long integer
value. The default unit of allocation shall be bytes. The optional string attribute
ovf:capacityAllocationUnits may be used to specify a particular unit of allocation. Values for
ovf:capacityAllocationUnits shall match the format for programmatic units defined in DSP0004
with the restriction that the base unit shall be "byte". 
````